# 03 – Feature Engineering

### Purpose
This notebook creates engineered features that enhance the predictive power and interpretability of the TED CAN dataset.
All features are based on insights from Notebook 01 (EDA) and domain knowledge of public procurement.

### Steps
- Load cleaned‑saved datasets
- Apply feature engineering pypline
- Save full datasets

--------------------
#### Imports & Setup & Dataset
-------------------

In [1]:
# ---------------------------------------------------------
# Import moduls
# ---------------------------------------------------------


# import standard modules
import pandas as pd
import numpy as np
from pathlib import Path

import sys
from pathlib import Path

In [2]:
# shut off some annoying warnings
import warnings

warnings.filterwarnings("ignore", message="A value is trying to be set on a copy")
warnings.filterwarnings("ignore", category=UserWarning)
warnings.filterwarnings("ignore", category=FutureWarning)

In [3]:
# ---------------------------------------------------------
# Setup style
# ---------------------------------------------------------

# show all columns
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# visualisation settings
pd.set_option('display.float_format', '{:,.2f}'.format)

In [5]:
# ---------------------------------------------------------
# Load scripts
# ----------------------------------------------------------

%load_ext autoreload
%autoreload 2

from pathlib import Path
import sys

# found min directory
PROJECT_ROOT = Path("..").resolve()

# maindirectory sys.path
if str(PROJECT_ROOT) not in sys.path:
    sys.path.insert(0, str(PROJECT_ROOT))

# import function from script
from my_scripts.feature_engineering import feature_engineering
from my_scripts.eda import overview

The autoreload extension is already loaded. To reload it, use:
  %reload_ext autoreload


In [6]:
# ---------------------------------------------------------
# Import dataset
# ---------------------------------------------------------

df = pd.read_pickle("../data/dataset_clean.pkl")

print("EU dataset:", df.shape)

EU dataset: (3521623, 27)


---------------------------------------------------------
### Apply feature engineering

-----------------------------

In [7]:
# apply funktion
df_fe = feature_engineering(df)


In [8]:
# shape of the dataset
print(df_fe.shape)

(3521623, 31)


In [9]:
# inspekt of the EU dataset
overview(df_fe)

,dtype,total,missing_n,missing_%,uniques_n,uniques
YEAR,int16,3521623,0,0.00,9,"[2008, 2009, 2010, 2011, 2012, 2013, 2014, 201..."
ID_TYPE,int8,3521623,0,0.00,7,"[3, 6, 18, 25, 21, 23, 22]"
CANCELLED,int8,3521623,0,0.00,2,"[0, 1]"
CORRECTIONS,int8,3521623,0,0.00,10,"[0, 1, 2, 3, 5, 4, 84, 7, 8, 6]"
CAE_TYPE,str,3521623,0,0.00,10,"[8, 3, 1, 6, R, N, 4, 5A, 5, Z]"
B_AWARDED_BY_CENTRAL_BODY,str,3521623,0,0.00,3,"[Unknown, Y, N]"
TYPE_OF_CONTRACT,str,3521623,0,0.00,3,"[W, U, S]"
B_DYN_PURCH_SYST,int64,3521623,0,0.00,2,"[0, 1]"
LOTS_NUMBER,float64,3521623,0,0.00,354,"[0.0, 3.0, 2.0, 1.0, 4.0, 10.0, 7.0, 13.0, 25...."
VALUE_EURO,float64,3521623,0,0.00,611809,"[512637.02, 5204457.56, 4924202.16, 1400000.0,..."


#### Notes: Feature Engineering Summary
Feature engineering enriches the cleaned dataset with additional analytical variables that improve modelling of tender competition, risk, and procedural behaviour.

1. Dataset Size After Features Engineering
  - Before: 3,521,623 rows × 27 columns
  - After: 3,521,623 rows × 31 columns

2. Created Features
- Date‑based features
  - AWARD_MONTH
  - AWARD_QUARTER
  - DAYS_TO_AWARD (time between dispatch and award)
These variables capture temporal patterns, seasonality, and procedural duration.

- CPV map with 12 group
  - "Construction & Building Works": 45–46
  - "Machinery & Industrial Equipment": 30–32
  - "Electrical, Optical & Precision Equipment": 33–34
  - "IT, Software & Telecommunications": 48–50
  - "Transport Services & Logistics": 60–63
  - "Business, Consulting & Administrative Services": 79–80
  - "Cleaning, Facility & Maintenance Services": 90–91
  - "Energy, Utilities & Environmental Services": 92–99
  - "Other"

- Boolean indicators (Y/N into 1/0)
  - B_ELECTRONIC_AUCTION
  - B_DYN_PURCH_SYST
  - B_ACCELERATED
  - B_AWARDED_TO_A_GROUP
  - B_CONTRACTOR_SME
All Y/N fields were mapped to binary numeric format.

- Missing‑information flags
  - VALUE_EURO_MISSING
  - AWARD_VALUE_EURO_MISSING
These indicators allow the model to capture informative missingness.

3. Missing numiric values replaced by the median.

4. Removed Raw Columns
After generating engineered features, the following original fields were dropped to avoid redundancy:
  - "DT_AWARD", "DT_DISPATCH"
  - "MAIN_CPV_CODE_GPA"


------------------
### Save engineered datasets

-------------------

In [10]:
df_fe.to_pickle("../data/dataset_fe.pkl")